### PR TITLE
Docker image for simple simulation

### PR DESCRIPTION
**Pull request by skim0119**
## Note

Here is the initial attempt to package everything within the docker. I checked the basic functionalities are correctly installed. The purpose is to provide "trial" playground to run simple models.

This build can be used to port `conda-forge` lateron.

## Feature

- Based on `Ubuntu`
- Install MPI, parallel-HDF5, Neuron, Neuroh5
- Python 3.10
- Jupyter Lab

## Future suggestions

- If we want users to interact with `jupyter-lab`, there might be a more optimized image to start.
- There is probably some more room for build optimization. The build takes quiet long.
- The image is about 4.62GB total. Again, there is probably some room to reduce space.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# Neuron (nrn)
+RUN pip install --no-cache-dir neuron
+
+# MiV Packages
+RUN mkdir -p /home/shared
+WORKDIR /home/shared
+
+RUN git clone https://github.com/GazzolaLab/MiV-Simulator \
+    && cd MiV-Simulator \
+    && poetry lock -n; poetry install
```
Comment (frthjf) [inline Dockerfile:86]: Why is the `poetry lock -n` necessary here? It seems to add quite a bit of overhead.

**Review by frthjf — COMMENTED** [review]
Running this locally, making neuroh5 fails due to a missing numpy dependency.

<details>
  <summary>Output log</summary>
  Step 24/35 : RUN cd neuroh5     && cmake . && make -j4
   ---> Running in a1c094a0d7b5
  -- The C compiler identification is GNU 11.2.0
  -- The CXX compiler identification is GNU 11.2.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Found HDF5: /opt/hdf5-1.12.1/build/lib/libhdf5.so;/usr/lib/x86_64-linux-gnu/libz.so;/usr/lib/x86_64-linux-gnu/libdl.a;/usr/lib/x86_64-linux-gnu/libm.so (found version "1.12.1")  
  found HDF5 version 1.12.1
  -- Found MPI_C: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so (found version "3.1") 
  CMake Warning (dev) at /usr/share/cmake-3.22/Modules/FindMPI.cmake:1470 (option):
    Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
    --help-policy CMP0077" for policy details.  Use the cmake_policy command to
    set the policy and suppress this warning.
  
    For compatibility with older versions of CMake, option is clearing the
    normal variable 'MPI_CXX_SKIP_MPICXX'.
  Call Stack (most recent call first):
    CMakeLists.txt:64 (find_package)
  This warning is for project developers.  Use -Wno-dev to suppress it.
  
  -- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1") 
  -- Found MPI: TRUE (found version "3.1")  
  found MPI version: 3.1
  MPI libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
  -- Found Python: /usr/bin/python3.10 (found suitable version "3.10.6", minimum required is "3") found components: Interpreter Development Development.Module Development.Embed 
  -- Python_EXECUTABLE="/usr/bin/python3.10"
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
  ModuleNotFoundError: No module named 'numpy'
  CMake Error at CMakeLists.txt:109 (message):
    NumPy headers not found
  
  
  found mpi4py include dir /usr/local/lib/python3.10/dist-packages/mpi4py/include
  -- Could NOT find JeMalloc (missing: JEMALLOC_LIBRARIES JEMALLOC_INCLUDE_DIR) 
  -- ParMETIS library: PARMETIS_LIBRARY-NOTFOUND
  -- Looking for pthread.h
  -- Looking for pthread.h - found
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
  -- Found Threads: TRUE  
  -- Could NOT find GTest (missing: GTEST_LIBRARY GTEST_INCLUDE_DIR GTEST_MAIN_LIBRARY) 
  CMake Warning at CMakeLists.txt:151 (message):
    gtest library not found, tests will not be run
  
  
  -- Install prefix: /usr/local
  -- Build type: RelWithDebInfo
  -- Building tests: ON
  -- Building documentation: 
  -- Building python bindings: ON
  -- Configuring incomplete, errors occurred!
  See also "/opt/neuroh5/CMakeFiles/CMakeOutput.log".
  The command '/bin/sh -c cd neuroh5     && cmake . && make -j4' returned a non-zero code: 1
</details>

Adding `RUN pip install numpy` fixes it, but technically `RUN pip install ./neuroh5` should pull that dependency automatically - not sure what is missing there.

Once the image is build, running `import miv_simulator` did not work for me. I think the issue is that poetry install by default creates a venv but the notebook later does run using the global python installation.

File: docs/tutorial/docker.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Docker Image
+
+> Note: The docker image is not optimized for high-performance computing. The image file only provides necessary installation for users to quickly try-run small model.
+
+We provide `Dockerfile` to quick-install docker image. The image contains necessary dependencies to run the simulation. We use `juypter-lab` as an recommended-interface.
+
+```bash
+cd MiV-Simulator                         # Change directory to repository
+docker build --tag miv_env:0.1 .         # Build image
```
[the remark this answers is not in the document]
Comment (skim0119) [reply docs/tutorial/docker.md:9]: It's probably equivalent as the help-string says `docker build [OPTIONS] PATH`, but I agree that your syntax is more readable. I'll check if the result is equivalent, and change the documentation

File: Dockerfile (hunk 1 above)
Comment (skim0119) [reply Dockerfile]: removed in [d32d694](https://github.com/GazzolaLab/MiV-Simulator/pull/24/commits/d32d69428c42a559f709694e60575a3468bc5b8d)

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+RUN pip install MiV-OS
```
[the remark this answers is not in the document]
Comment (skim0119) [reply Dockerfile]: removed in [d32d694](https://github.com/GazzolaLab/MiV-Simulator/pull/24/commits/d32d69428c42a559f709694e60575a3468bc5b8d)

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+RUN pip install ./neuroh5
+RUN cd neuroh5 \
+    && cmake . && make -j4
+RUN echo "export PATH=/opt/neuroh5/bin:$PATH" >> ~/.bashrc
+ENV PATH /opt/neuroh5/bin:$PATH
+RUN which h5copy \
+    && which neurotrees_import
+
+# Neuron (nrn)
+RUN pip install --no-cache-dir neuron
```
[the remark this answers is not in the document]
Comment (skim0119) [reply Dockerfile]: [d32d694](https://github.com/GazzolaLab/MiV-Simulator/pull/24/commits/d32d69428c42a559f709694e60575a3468bc5b8d)

File: docs/tutorial/docker.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (skim0119) [reply docs/tutorial/docker.md]: [d32d694](https://github.com/GazzolaLab/MiV-Simulator/pull/24/commits/d32d69428c42a559f709694e60575a3468bc5b8d)

[the remark this answers is not in the document]
Comment (frthjf) [reply docs/tutorial/docker.md]: Ah yep, the other version for fine as well, it's just easy to miss the `.` when copy-pasting.

File: Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (skim0119) [reply Dockerfile]: I'm adding miv-os back to install graphic packages in first example case.

**Comment by skim0119** [comment]
I checked the scripts are running okay. I'm merging this PR (squashed).

**Pull request merged**